### PR TITLE
Merge changes from regression related tools

### LINKED ISSUE #1415: Consolidate multiple bot changes into a single change for regression bugs
If you file a bug that's regressed by another bug but don't correctly fill in all the fields, you then get at least three separate updates from the release management bot.

For example, bug 1766234 just got multiple updates:

* Release status flags
* Needinfo of author
* Regression range set

This is causing more visual noise and unnecessary email spam spread over a longer period. Please could we consider somehow consolidating these changes into one?

### PR DESCRIPTION
**Pull request by suhaibmujahid**
Resolves #1415

**Note:** reviewing commit by commit will be clearer

## Checklist

- [x] Type annotations added to new functions
- [x] Docs added to functions touched in main classes
- [x] Dry-run produced the expected results
- [x] The [`to-be-announced`](https://github.com/mozilla/relman-auto-nag/labels/to-be-announced) tag added if this is worth announcing

### CODE REVIEW

**Review by calixteman — APPROVED** [review]
It looks good overall.
I don't see anything obviously wrong so r=me.

File: auto_nag/bzcleaner.py [hunk 1]
```diff
@@ -581,37 +582,71 @@ def autofix(self, bugs):
...
+        tool_name: str,
+        new_changes: Dict[str, dict],
+        no_bugmail: bool = False,
+        is_dryrun: bool = True,
+        db_extra: Dict[str, str] = None,
+    ) -> None:
+        """Apply changes on Bugzilla
+
+        Args:
+            tool_name: the name of the tool that are performing the changes.
```
Comment (Archaeopteryx) [inline auto_nag/bzcleaner.py:611]: `are` => `is`

File: auto_nag/bzcleaner.py [hunk 2]
```diff
@@ -581,37 +582,71 @@ def autofix(self, bugs):
...
+            new_changes: the changes that will be performed. The dictatory key
```
Comment (Archaeopteryx) [inline auto_nag/bzcleaner.py:612]: typo

File: auto_nag/bzcleaner.py [hunk 3]
```diff
@@ -581,37 +582,71 @@ def autofix(self, bugs):
...
+                should be the bug ID.
+            no_bugmail: If True, an account that doesn't trigger bugmail will be
+                used to apply the changes.
+            is_dryrun: If True, no changes will be applied. Instead, the
+                proposed changes will be logged.
+            db_extra: extra data to be passed to the DB. The dictatory key
```
Comment (Archaeopteryx) [inline auto_nag/bzcleaner.py:618]: typo

File: auto_nag/multi_autofixers.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+class MultiAutoFixers:
+    "Merge changes from multiple tools and apply them to Bugzilla at once"
+
+    def __init__(
+        self, *tools: BzCleaner, **merge_functions: Callable[[ToolsChanges], dict]
+    ):
+        """Constructor
+
+        Args:
+            *tools: tools to merged their changes.
```
Comment (Archaeopteryx) [inline auto_nag/multi_autofixers.py:60]: typo + next line

File: auto_nag/multi_autofixers.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    def description(self):
+        """Return a description of the tool"""
+        return "Grouped changes for the following tools: " + self.name()
+
+    def name(self):
+        """Return the name of the tool"""
+        return utils.english_list([tool.name() for tool in self.tools])
+
+    def get_args_parser(self):
+        """Get the argumends from the command line"""
```
Comment (Archaeopteryx) [inline auto_nag/multi_autofixers.py:85]: Typo

**Review by Archaeopteryx — APPROVED** [review]

File: auto_nag/multi_autofixers.py (hunk 5 above)
Comment (suhaibmujahid) [reply auto_nag/multi_autofixers.py:85]: Nice catch. This typo has two clones that have been living in the repo for 4 years :)

Fixed in b84ea58cdf07dce2eb0abc5604d8f114cb6226d8

File: auto_nag/multi_autofixers.py (hunk 4 above)
Comment (suhaibmujahid) [reply auto_nag/multi_autofixers.py:60]: Fixed in d5ee660b1d9e6a3a04e603b54bfd570bc181c32c

File: auto_nag/tests/test_multi_autofixers.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+        raise UnexpectedToolsError(list(tools))
+
+
+class MissingMergeFunctionMockup(MultiAutoFixers):
+    def __init__(self):
+        super().__init__(ToolThreeMockup(), ToolFourMockup())
+
+
+class UnsportedToolInMergeFunctionMockup(MultiAutoFixers):
```
[the remark this answers is not in the document]
Comment (suhaibmujahid) [reply auto_nag/tests/test_multi_autofixers.py:103]: Fixed in d5ee660b1d9e6a3a04e603b54bfd570bc181c32c

File: auto_nag/bzcleaner.py (hunk 1 above)
Comment (suhaibmujahid) [reply auto_nag/bzcleaner.py:611]: Fixed in 762bc5f791659ecc302676ffe54e9d575e0b5902

File: auto_nag/bzcleaner.py (hunk 2 above)
Comment (suhaibmujahid) [reply auto_nag/bzcleaner.py:612]: Fixed in 762bc5f791659ecc302676ffe54e9d575e0b5902

File: auto_nag/bzcleaner.py (hunk 3 above)
Comment (suhaibmujahid) [reply auto_nag/bzcleaner.py:618]: Fixed in 762bc5f791659ecc302676ffe54e9d575e0b5902

**Review by Archaeopteryx — APPROVED** [review]

**Comment by suhaibmujahid** [comment]
# Dry-run

## Current seperated changes (using bug [1782311](https://bugzilla.mozilla.org/show_bug.cgi?id=1782311) as an example)

### [Comment 7](https://bugzilla.mozilla.org/show_bug.cgi?id=1782311#c7)
Set release status flags based on info from the regressing bug [1709405](https://bugzilla.mozilla.org/show_bug.cgi?id=1709405)

| Field                    | Value                             |
| ------------------------ | --------------------------------- |
| cf_status_firefox103     | affected                          |
| cf_status_firefox104     | affected                          |
| cf_status_firefox_esr102 | affected                          |
| cf_status_firefox_esr91  | affected                          |

### [Comment 8](https://bugzilla.mozilla.org/show_bug.cgi?id=1782311#c8)

:daisuke, since you are the author of the regressor, bug [1709405](https://bugzilla.mozilla.org/show_bug.cgi?id=1709405), could you take a look?
For more information, please visit [auto_nag documentation](https://wiki.mozilla.org/Release_Management/autonag#needinfo_regression_author.py).

| Field                    | Value                             |
| ------------------------ | --------------------------------- |
| flags                    | needinfo?(daisuke@xxxx.xxx) |

## Aggregated changes
Set release status flags based on info from the regressing bug [1709405](https://bugzilla.mozilla.org/show_bug.cgi?id=1709405)

:daisuke, since you are the author of the regressor, bug [1709405](https://bugzilla.mozilla.org/show_bug.cgi?id=1709405), could you take a look?
For more information, please visit [auto_nag documentation](https://wiki.mozilla.org/Release_Management/autonag#needinfo_regression_author.py).

| Field                    | Value                             |
| ------------------------ | --------------------------------- |
| flags                    | needinfo?(daisuke@xxxx.xxx) |
| cf_status_firefox103     | affected                          |
| cf_status_firefox104     | affected                          |
| cf_status_firefox_esr102 | affected                          |
| cf_status_firefox_esr91  | affected                          |

**Pull request merged**